src/Select.purs

```diff
@@ -115,47 +115,49 @@ type Input st =
   | st
   }
 
-type Spec st query act ps msg m =
-  { -- usual Halogen component spec 
-    render 
-      :: State st 
-      -> H.ComponentHTML (Action act) ps m
-    
+type Spec st query action slots input msg m =
+  { -- usual Halogen component spec
+    render
+      :: State st
+      -> H.ComponentHTML (Action action) slots m
+
     -- handle additional actions provided to the component
-  , handleAction 
-      :: act
-      -> H.HalogenM (State st) (Action act) ps msg m Unit
+  , handleAction
+      :: action
+      -> H.HalogenM (State st) (Action action) slots msg m Unit
 
     -- handle additional queries provided to the component
-  , handleQuery 
+  , handleQuery
       :: forall a
-       . query a 
-      -> H.HalogenM (State st) (Action act) ps msg m (Maybe a)
+       . query a
+      -> H.HalogenM (State st) (Action action) slots msg m (Maybe a)
 
     -- handle messages emitted by the component; provide H.raise to simply
     -- raise the Select messages to the parent.
-  , handleMessage 
+  , handleMessage
       :: Message
-      -> H.HalogenM (State st) (Action act) ps msg m Unit
+      -> H.HalogenM (State st) (Action action) slots msg m Unit
 
     -- optionally handle input on parent re-renders
-  , receive 
-      :: Input st 
-      -> Maybe (Action act)
+  , receive
+      :: input
+      -> Maybe (Action action)
 
     -- perform some action when the component initializes.
-  , initialize 
-      :: Maybe (Action act)
+  , initialize
+      :: Maybe (Action action)
 
     -- optionally perform some action on initialization. disabled by default.
   , finalize
-      :: Maybe (Action act)
+      :: Maybe (Action action)
```

Comment:
If we change this record to be this instead...

```purescript
type Spec st query action slots input msg m =
  { -- usual Halogen component spec
    render :: State st -> H.ComponentHTML (Action action) slots m

    -- handle additional actions provided to the component
  , handleAction :: action -> H.HalogenM (State st) (Action action) slots msg m Unit

    -- handle additional queries provided to the component
  , handleQuery :: forall a. query a -> H.HalogenM (State st) (Action action) slots msg m (Maybe a)

    -- handle messages emitted by the component; provide H.raise to simply
    -- raise the Select messages to the parent.
  , handleMessage :: Message -> H.HalogenM (State st) (Action action) slots msg m Unit

    -- optionally handle input on parent re-renders
  , receive :: input -> Maybe action

    -- perform some action when the component initializes.
  , initialize :: Maybe action

    -- optionally perform some action on initialization. disabled by default.
  , finalize :: Maybe action
  }
```

...then we can change the `Initialize` data constructor to be

```purescript
data Action 
  = ...
  | Initialize (Maybe action)
```

and the `handleAction` for initialization to be...

```purescript
  Initialize -> do
    ...
    for_ mbAction (Action >>> handle)
```

and the component spec to be:

```purescript
component mkInput spec = H.mkComponent
  { initialState: initialState <<< mkInput
  , render: spec.render
  , eval: H.mkEval $ H.defaultEval
      { handleQuery = handleQuery spec.handleQuery
      , handleAction = handleAction spec.handleAction spec.handleMessage
      , initialize = Just (Initialize spec.initialize)
      , receive = map Action <<< spec.receive
      , finalize = map Action spec.finalize
      }
  }
  where
  initialState :: Input st -> State st
  initialState = Builder.build pipeline
    ...
```

This fixes your note in which these map to `Action action` instead of just to a user-provided action.

Reply:
Great! I was wanting to make those changes here, too.

> Also, I would prefer to name the input -> Input st something other than initialState and keep the Input st -> State st function named initialState. In this example I've renamed the new input -> Input st function simply mkInput, but you may have a better name for it.

Yeah, I can do that. For a better name, what about `incompleteState`? `mkInput` sounds good, but is `Input` really the best name to give for that now? It's no longer the input value to the component.

Reply:
@davezuch do you have any opinions on the names here?

Reply:
It's a bit confusing because we have various "inputs" going on:

1. The Select component requires a standard Halogen input type, currently called `Input st`
2. An extended Select component can expose a different input type to its users, currently represented by the type parameter `input`
3. The Select component produces its initial state from the `Input st` type as in usual Halogen, a function currently called `initialState`
4. An extended Select component requires a function to transform the `input` type into Select's `Input st` type, currently called `mkInput`.

So there's `Input` (for Select) and `input` (for extended Select); and there's `initialState :: Input -> State` (for Select) and `mkInput :: input -> Input st` (for extended Select).

Does this seem understandable, or does this bit of text bring to mind more apt names & types?

Reply:
Here's my proposal. These numbers don't correspond to Thomas' list above. They simply indicate how many things need to change:

1. Rename `Input st` to `IntermediateState st` as that's really what it is: an intermediate state value that is used to produce the component's actual initial state
2. Rename `mkInput` to `mkIntermediateState` to match the type change
3. Rename the Select component's `initialState` to `completeInitialState` to make the composition of these two functions sensible: `completeInitialState <<< mkIntermediateState`

As I was typing this comment, I realized it might make sense to rename `InputType` to `ComponentType` 

```purescript
{- old version:
data InputType = Text | Toggle                                       -}
data ComponentType = Text | Toggle

-- old version:
-- type Input st =
type IntermediateState st =
-- old version:
--{ inputType :: InputType
  { componentType :: ComponentType
  , search :: Maybe String
  , debounceTime :: Maybe Milliseconds
  , getItemCount :: {| st } -> Int
  | st
  }

-- proposed version only shown here
component :: MassivelyLongTypeSignatureIgnoredHere
component mkIntermediateState spec = H.mkComponent
  { initialState: completeInitialState <<< mkIntermediateState
  , -- render code 
  , -- eval code
  }
  where
  completeInitialState :: Input st -> State st
  completeInitialState = Builder.build pipeline
    -- rest of code
```

Reply:
This conversation is relevant over in Formless:
https://github.com/thomashoneyman/purescript-halogen-formless/pull/47#issuecomment-495710350

I'm hesitant to move to intermediateState because without extension the `Input` type as it stands really _is_ the component input, the same way the `Message` type really _is_ the message output without extension. I can easily see cases in which you might not want to change the Select input type and keep it as-is.

Reply:
I respond to both the "`input` is the component input when no extension occurs" and "`message` is the output when no extension occurs" concerns here:
- input concern: https://github.com/thomashoneyman/purescript-halogen-formless/pull/47#issuecomment-495744919
- message concern: https://github.com/thomashoneyman/purescript-halogen-formless/pull/47#issuecomment-495747988

Reply:
Another idea that came to mind: we could remove the `Input st` type completely and change the type of `mkInput` to inline the definition of the `Input st` type.
```purescript
-- current
mkInput :: input -> Input st

-- proposal
mkInput :: input -> 
  { inputType :: InputType
  , search :: Maybe String
  , debounceTime :: Maybe Milliseconds
  , getItemCount :: {| st } -> Int
  | st
  }
```

Reply:
And yet another idea (sorry for the multiple posts). @thomashoneyman Perhaps we can both get behind this one:
```purescript
type MakeInput input st =
  input -> { inputType :: InputType
           , search :: Maybe String
           , debounceTime :: Maybe Milliseconds
           , getItemCount :: {| st } -> Int
           | st
           }

component :: MakeInput input st -> Spec -> Component -- rest of type signature
```